src/python/pants/backend/python/rules/python_test_runner_integration_test.py

```diff
@@ -1,266 +1,376 @@
+
+  source_root = "tests/python/pants_test"
+  good_source = FileContent(path="test_good.py", content=b"def test():\n  pass\n")
+  bad_source = FileContent(path="test_bad.py", content=b"def test():\n  assert False\n")
+  py3_only_source = FileContent(path="test_py3.py", content=b"def test() -> None:\n  pass\n")
+  library_source = FileContent(path="library.py", content=b"def add_two(x):\n  return x + 2\n")
+
+  create_python_library = partialmethod(
+    TestBase.create_library, path=source_root, target_type="python_library",
+  )
+
+  def write_file(self, file_content: FileContent) -> None:
+    self.create_file(
+      relpath=str(PurePath(self.source_root, file_content.path)),
+      contents=file_content.content.decode(),
     )
 
-    for line_number, (want_line, got_line) in enumerate(zip(want_lines, got_lines)):
-      want_parts = want_line.split('SOME_TEXT')
-      if len(want_parts) == 1:
-        self.assertEqual(want_line, got_line,
-                         "Line {} wrong: want '{}', got '{}'"
-                         .format(line_number, want_line, got_line))
-      elif len(want_parts) == 2:
-        self.assertTrue(got_line.startswith(want_parts[0]), f'Line {line_number} Want "{got_line}" to start with "{want_parts[0]}"')
-        self.assertTrue(got_line.endswith(want_parts[1]), f'Line {line_number} Want "{got_line}" to end with "{want_parts[1]}"')
-
-  def run_passing_pants_test(self, trailing_args):
-    pants_run = self.run_pants(trailing_args)
-    self.assert_success(pants_run)
-    return pants_run
-
-  def run_failing_pants_test(self, trailing_args):
-    pants_run = self.run_pants(trailing_args)
-    self.assert_failure(pants_run)
-    self.assertIn('Tests failed\n', pants_run.stderr_data)
-    return pants_run
-
-  def test_passing_test(self):
-    pants_run = self.run_passing_pants_test([
-      'testprojects/tests/python/pants/dummies:passing_target',
-    ])
-    self.assert_fuzzy_string_match(
-      pants_run.stdout_data,
-      dedent("""\
-        testprojects/tests/python/pants/dummies:passing_target stdout:
-        ============================= test session starts ==============================
-        platform SOME_TEXT
-        rootdir: SOME_TEXT
-        plugins: SOME_TEXT
-        collected 1 item
-
-        pants/dummies/test_pass.py .                                             [100%]
-
-        ============================== 1 passed in SOME_TEXT ===============================
-
-
-        testprojects/tests/python/pants/dummies:passing_target                          .....   SUCCESS
-        """),
+  def create_basic_library(self) -> None:
+    self.create_python_library(name="library", sources=[self.library_source.path])
+    self.write_file(self.library_source)
+
+  def create_python_test_target(
+    self,
+    source_files: List[FileContent],
+    *,
+    dependencies: Optional[List[str]] = None,
+    interpreter_constraints: Optional[str] = None,
+  ) -> None:
+    self.add_to_build_file(
```

Comment:
We can't use `self.create_library` because it wouldn't actually add the field `compatibility` to the BUILD file.